src/services/ssdi.ts

```diff
@@ -1,197 +1,632 @@
+  defaultValues?: string[] | number[] | boolean[] | Record<string, any>[];
+  conditions?: string[];
 }
 
-// Caches to store the parsed query files and their metadata
-const queryFileCache: Map<string, QueryFile> = new Map();
-const queryDataCache: Map<string, CachedQuery> = new Map();
+interface Filters {
+  [key: string]: Filter;
+}
 
-// Helper function to read name, description, and default output name from a file
-const readNameAndDescriptionFromFile = (
-  filePath: string,
-): {
-  name: string;
-  description: string;
-  defaultOutputName: string;
-} => {
-  const fileContents = fs.readFileSync(filePath, 'utf8');
-  const nameMatch = fileContents.match(/@name:\s*(.*)/);
-  const descriptionMatch = fileContents.match(/@description:\s*(.*)/);
-  const defaultOutputNameMatch = fileContents.match(/@defaultOutputName:\s*(.*)/);
-  const fileName = path.basename(filePath, path.extname(filePath));
-  return {
-    name: nameMatch ? nameMatch[1].trim() : fileName,
-    description: descriptionMatch ? descriptionMatch[1].trim() : 'No description available',
-    defaultOutputName: defaultOutputNameMatch ? defaultOutputNameMatch[1].trim() : fileName,
-  };
+interface FilterValues {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [key: string]: string[] | number[] | boolean[] | Record<string, any>[];
+}
+
+interface CachedFilters {
+  filters: Filters;
+  supportsSorting: boolean;
+  supportsPagination: boolean;
+}
+
+// Base directory for file operations
+const BASE_DIRECTORY = path.resolve(process.cwd(), '/src/queries/');
+
+// Cache to store parsed JSON headers and queries
+const cache: Map<string, CachedFile> = new Map();
+
+// Sanitize the filename
+const sanitizeFilename = (filename: string): string => path.normalize(filename).replace(/[^a-zA-Z0-9-_]/g, '_');
+
+// Check if the user has access to a specific folders
+const checkFolderPermissions = async (
+  user,
+  scriptPath: string,
+) => (new Generic(user)).checkPermissions(
+  scriptPath,
+  ['dataRequests/ohs', 'dataRequests/internal'],
+  'ssdi_restricted',
+);
+
+// Safe file path resolution to prevent directory traversal attacks
+const safeResolvePath = (inputPath: string): string => {
+  const resolvedPath = path.resolve(BASE_DIRECTORY, inputPath);
+
+  if (!resolvedPath.startsWith(BASE_DIRECTORY)) {
+    throw new Error('Attempted to access a file outside the allowed directory.');
+  }
+
+  return resolvedPath;
 };
 
-// Helper function to list all query files with name and description
-const listQueryFiles = (directory: string): QueryFile[] => {
-  const files = fs.readdirSync(directory);
-  return files.map((file) => {
-    const filePath = path.join(directory, file);
-    if (queryFileCache.has(filePath)) {
-      return queryFileCache.get(filePath) as QueryFile;
+// Basic JSON validation function
+// TODO: make this more verbose validation
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const isValidJsonHeader = (json: any): boolean => (
```

Comment:
/**
 * Refactored `isValidJsonHeader` to use Zod for validation based on `HeaderStructure` interface.
 * 
 * To automatically convert TypeScript interfaces to Zod schemas, we can use `ts-to-zod`:
 * 
 * 1. Install `zod` and `ts-to-zod`:
 * 
 * ```bash
 * npm install zod ts-to-zod
 * ```
 * 
 * 2. Define your existing TypeScript interfaces for `HeaderFilter` and `HeaderStructure`.
 * 
 * 3. Use `ts-to-zod` to generate the Zod schema from the TypeScript interface.
 * 
 * Example:
 * 
 * ```ts
 * import { z } from 'zod';
 * import { zodToSchema } from 'ts-to-zod';
 * 
 * // Your existing interfaces
 * interface HeaderFilter {
 *   name: string;
 *   type: string;
 *   display: string;
 *   description: string;
 *   supportsExclusion?: boolean;
 *   supportsFuzzyMatch?: boolean;
 *   options?: {
 *     query?: {
 *       sqlQuery: string;
 *       column: string;
 *     };
 *     staticValues?: string[] | number[] | boolean[] | Record<string, any>[];
 *     defaultValues?: string[] | number[] | boolean[] | Record<string, any>[];
 *   };
 * }
 * 
 * interface HeaderStructure {
 *   name: string;
 *   description: {
 *     standard: string;
 *     technical: string;
 *   };
 *   output: {
 *     defaultName: string;
 *     schema: Array<{
 *       columnName: string;
 *       type: string | number | boolean | Record<string, any>;
 *       nullable: boolean;
 *       description: string;
 *     }>;
 *     sorting?: {
 *       default: Array<{
 *         name: string;
 *         order: 'ASC' | 'DESC';
 *       }>;
 *       supportsCustomSorting?: boolean;
 *     };
 *     supportsPagination?: boolean;
 *     multipleDataSets?: Array<{
 *       name: string;
 *       defaultName: string;
 *       description: string;
 *       schema: Array<{
 *         columnName: string;
 *         type: string | number | boolean | Record<string, any>;
 *         nullable: boolean;
 *         description: string;
 *       }>;
 *       sorting?: {
 *         default: Array<{
 *           name: string;
 *           order: 'ASC' | 'DESC';
 *         }>;
 *         supportsCustomSorting?: boolean;
 *       };
 *       supportsPagination?: boolean;
 *     }>;
 *   };
 *   filters: Array<HeaderFilter>;
 * }
 * 
 * // Use ts-to-zod to generate the schema from the interface
 * const HeaderFilterSchema = zodToSchema<HeaderFilter>();
 * const HeaderStructureSchema = zodToSchema<HeaderStructure>();
 * 
 * // Refactor isValidJsonHeader to use Zod for validation
 * const isValidJsonHeader = (json: unknown): boolean => {
 *   try {
 *     HeaderStructureSchema.parse(json); // Validate json against the generated Zod schema
 *     return true;
 *   } catch (e) {
 *     console.error(e.errors); // Log detailed validation errors if needed
 *     return false;
 *   }
 * };
 * ```
 * 
 * With this setup, the `isValidJsonHeader` function now uses the automatically generated Zod schema
 * based on the TypeScript interface. You can maintain your validation logic in one place by leveraging
 * TypeScript interfaces and Zod at runtime.
 */